File: mcp/message-security.ts
// Verifies message provenance before boss content enters MCP model context.
// Exports verifiable message types, recursive ES256 checks, and assurance labels.
// Depends on Web Crypto and strict JSON/base64url parsing only.

const SIGNATURE_SCHEME = 'JWS-ES256';
const JWS_TYPE = 'hiboss-message+jws';
const MESSAGE_PURPOSE = 'hiboss.boss-message';

export interface VerifiableMessage {
  readonly id: string;
  readonly body: string;
  readonly direction: string;
  readonly reply_to?: string | null;
  readonly metadata?: Record<string, unknown> | null;
  readonly replies?: VerifiableMessage[];
}

export type MessageAssurance =
  | { readonly kind: 'verified'; readonly source: string; readonly keyId: string }
  | { readonly kind: 'attributed'; readonly source: string; readonly status: string }
  | { readonly kind: 'not_applicable' };

interface ParsedJws {
  readonly header: Record<string, unknown>;
  readonly payload: Record<string, unknown>;
  readonly input: Uint8Array;
  readonly signature: Uint8Array;
}

export async function verifyMessage(message: VerifiableMessage): Promise<MessageAssurance> {
  const assurance = await verifySingle(message);
  for (const reply of message.replies ?? []) await verifyMessage(reply);
  return assurance;
}

export async function verifyMessages(messages: VerifiableMessage[]): Promise<void> {
  for (const message of messages) await verifyMessage(message);
}

export function assuranceLabel(assurance: MessageAssurance): string {
  if (assurance.kind === 'verified') return `${assurance.source}/verified`;
  if (assurance.kind === 'attributed') return `${assurance.source}/${assurance.status}`;
  return 'agent';
}

async function verifySingle(message: VerifiableMessage): Promise<MessageAssurance> {
  if (message.direction !== 'boss_to_agent') return { kind: 'not_applicable' };
  const metadata = message.metadata;
  if (!metadata) throw failure(message, 'missing metadata');
  const provenance = asRecord(metadata.provenance);
  if (!provenance) throw failure(message, 'missing provenance');
  if (provenance.version !== 1) throw failure(message, 'unsupported provenance version');
  const source = requiredString(message, provenance, 'source');
  if (metadata.source !== source) throw failure(message, 'source attribution mismatch');
  const signature = asRecord(provenance.signature);
  if (!signature) throw failure(message, 'missing signature state');
  return signature.scheme === undefined
    ? verifyAttribution(message, signature, source)
    : verifySignature(message, provenance, signature, source);
}

function verifyAttribution(
  message: VerifiableMessage,
  signature: Record<string, unknown>,
  source: string,
): MessageAssurance {
  const status = requiredString(message, signature, 'status');
  const expected = source === 'discord' || source === 'telegram'
    ? 'unsupported'
    : source === 'api'
      ? 'not_configured'
      : source === 'system' ? 'not_applicable' : null;
  if (!expected) throw failure(message, 'unsigned native source');
  if (status !== expected) throw failure(message, 'invalid signature status for source');
  return { kind: 'attributed', source, status };
}

async function verifySignature(
  message: VerifiableMessage,
  provenance: Record<string, unknown>,
  signature: Record<string, unknown>,
  source: string,
): Promise<MessageAssurance> {
  if (source !== 'ios' && source !== 'macos') throw failure(message, 'signed source is not native');
  if (signature.scheme !== SIGNATURE_SCHEME) throw failure(message, 'unsupported signature scheme');
  const keyId = requiredString(message, signature, 'key_id');
  const publicKey = decodeRequired(message, signature, 'public_key');
  const derivedKeyId = new Uint8Array(await crypto.subtle.digest('SHA-256', arrayBuffer(publicKey)));
  if (!equalBytes(decode(message, keyId), derivedKeyId)) {
    throw failure(message, 'signing key identifier mismatch');
  }
  const compact = requiredString(message, signature, 'signed_message');
  const parsed = parseJws(message, compact);
  validateHeader(message, parsed.header, keyId);
  const key = await importPublicKey(message, publicKey);
  const valid = await crypto.subtle.verify(
    { name: 'ECDSA', hash: 'SHA-256' }, key,
    arrayBuffer(parsed.signature), arrayBuffer(parsed.input),
  );
  if (!valid) throw failure(message, 'invalid ES256 signature');
  validatePayload(message, provenance, parsed.payload);
  return { kind: 'verified', source, keyId };
}

function parseJws(message: VerifiableMessage, compact: string): ParsedJws {
  const segments = compact.split('.');
  if (segments.length !== 3 || segments.some((segment) => !segment)) {
    throw failure(message, 'malformed compact JWS');
  }
  const [encodedHeader, encodedPayload, encodedSignature] = segments as [string, string, string];
  const signature = decode(message, encodedSignature);
  if (signature.length !== 64) throw failure(message, 'invalid ES256 signature length');
  return {
    header: decodeJson(message, encodedHeader, 'header'),
    payload: decodeJson(message, encodedPayload, 'payload'),
    input: new TextEncoder().encode(`${encodedHeader}.${encodedPayload}`),
    signature,
  };
}

function validateHeader(
  message: VerifiableMessage,
  header: Record<string, unknown>,
  keyId: string,
): void {
  if (header.alg !== 'ES256' || header.typ !== JWS_TYPE || header.kid !== keyId) {
    throw failure(message, 'JWS header mismatch');
  }
}

function validatePayload(
  message: VerifiableMessage,
  provenance: Record<string, unknown>,
  payload: Record<string, unknown>,
): void {
  const action = asRecord(payload.action);
  if (payload.version !== 1 || payload.purpose !== MESSAGE_PURPOSE || action?.kind !== 'reply') {
    throw failure(message, 'signed payload contract mismatch');
  }
  if (typeof payload.message_id !== 'string' || !payload.message_id
    || typeof payload.issued_at !== 'number' || !Number.isInteger(payload.issued_at)) {
    throw failure(message, 'invalid signed message identity');
  }
  if (payload.body !== message.body) throw failure(message, 'signed body mismatch');
  if (action.message_id !== message.reply_to) throw failure(message, 'signed reply target mismatch');
  const actor = asRecord(provenance.actor);
  if (payload.boss_id !== actor?.id) throw failure(message, 'signed boss identity mismatch');
}

async function importPublicKey(message: VerifiableMessage, value: Uint8Array): Promise<CryptoKey> {
  if (value.length !== 65) throw failure(message, 'invalid P-256 public key length');
  try {
    return await crypto.subtle.importKey(
      'raw', arrayBuffer(value), { name: 'ECDSA', namedCurve: 'P-256' }, false, ['verify'],
    );
  } catch {
    throw failure(message, 'invalid P-256 public key');
  }
}

function decodeJson(
  message: VerifiableMessage,
  encoded: string,
  part: string,
): Record<string, unknown> {
  try {
    const value: unknown = JSON.parse(new TextDecoder().decode(decode(message, encoded)));
    const record = asRecord(value);
    if (!record) throw new Error('not an object');
    return record;
  } catch {
    throw failure(message, `invalid JWS ${part}`);
  }
}

function decodeRequired(
  message: VerifiableMessage,
  record: Record<string, unknown>,
  key: string,
): Uint8Array {
  return decode(message, requiredString(message, record, key));
}

function decode(message: VerifiableMessage, value: string): Uint8Array {
  if (!/^[A-Za-z0-9_-]+$/.test(value) || value.length % 4 === 1) {
    throw failure(message, 'invalid base64url encoding');
  }
  const bytes = Uint8Array.from(Buffer.from(value, 'base64url'));
  if (Buffer.from(bytes).toString('base64url') !== value) {
    throw failure(message, 'non-canonical base64url encoding');
  }
  return bytes;
}

function requiredString(
  message: VerifiableMessage,
  record: Record<string, unknown>,
  key: string,
): string {
  const value = record[key];
  if (typeof value !== 'string' || !value) throw failure(message, `missing ${key}`);
  return value;
}

function asRecord(value: unknown): Record<string, unknown> | null {
  return value && typeof value === 'object' && !Array.isArray(value)
    ? value as Record<string, unknown>
    : null;
}

function arrayBuffer(value: Uint8Array): ArrayBuffer {
  return value.slice().buffer as ArrayBuffer;
}

function equalBytes(left: Uint8Array, right: Uint8Array): boolean {
  return left.length === right.length && left.every((byte, index) => byte === right[index]);
}

function failure(message: VerifiableMessage, detail: string): Error {
  return new Error(`message ${message.id} failed provenance verification: ${detail}`);
}

File: cli/src/message_security.rs
// Purpose: Verify boss-message provenance before message bodies reach an AI agent.
// Exports signature enforcement, assurance labels, and recursive message checks.
// Dependencies: API Message types, serde JSON, and ring ES256 verification.

use crate::types::Message;
use ring::digest::{digest, SHA256};
use ring::signature::{UnparsedPublicKey, ECDSA_P256_SHA256_FIXED};
use serde::Deserialize;
use serde_json::{Map, Value};
use std::error::Error;
use std::fmt::{Display, Formatter};

const SIGNATURE_SCHEME: &str = "JWS-ES256";
const JWS_TYPE: &str = "hiboss-message+jws";
const MESSAGE_PURPOSE: &str = "hiboss.boss-message";

#[derive(Debug, PartialEq, Eq)]
pub enum MessageAssurance {
    Verified { source: String, key_id: String },
    Attributed { source: String, status: String },
    NotApplicable,
}

#[derive(Debug)]
pub struct MessageVerificationError {
    message_id: String,
    detail: String,
}

impl Display for MessageVerificationError {
    fn fmt(&self, formatter: &mut Formatter<'_>) -> std::fmt::Result {
        write!(formatter, "message {} failed provenance verification: {}", self.message_id, self.detail)
    }
}

impl Error for MessageVerificationError {}

#[derive(Deserialize)]
struct JwsHeader {
    alg: String,
    kid: String,
    typ: String,
}

#[derive(Deserialize)]
struct SignedAction {
    kind: String,
    message_id: String,
}

#[derive(Deserialize)]
struct SignedPayload {
    version: u8,
    purpose: String,
    message_id: String,
    issued_at: i64,
    boss_id: String,
    action: SignedAction,
    body: String,
}

pub fn verify_message(message: &Message) -> Result<MessageAssurance, MessageVerificationError> {
    let assurance = verify_single(message)?;
    for reply in message.replies.as_deref().unwrap_or_default() {
        verify_message(reply)?;
    }
    Ok(assurance)
}

pub fn verify_messages(messages: &[Message]) -> Result<(), MessageVerificationError> {
    for message in messages {
        verify_message(message)?;
    }
    Ok(())
}

pub fn parse_verified_message(value: &str) -> Result<Message, Box<dyn Error>> {
    let message = serde_json::from_str::<Message>(value)?;
    verify_message(&message)?;
    Ok(message)
}

pub fn assurance_label(message: &Message) -> String {
    match verify_single(message) {
        Ok(MessageAssurance::Verified { source, .. }) => format!("{source}/verified"),
        Ok(MessageAssurance::Attributed { source, status }) => format!("{source}/{status}"),
        Ok(MessageAssurance::NotApplicable) => "agent".to_owned(),
        Err(_) => "invalid".to_owned(),
    }
}

fn verify_single(message: &Message) -> Result<MessageAssurance, MessageVerificationError> {
    if message.direction.as_deref() != Some("boss_to_agent") {
        return Ok(MessageAssurance::NotApplicable);
    }
    let metadata = message.metadata.as_ref().ok_or_else(|| fail(message, "missing metadata"))?;
    let provenance = object(metadata.get("provenance"))
        .ok_or_else(|| fail(message, "missing provenance"))?;
    if provenance.get("version").and_then(Value::as_u64) != Some(1) {
        return Err(fail(message, "unsupported provenance version"));
    }
    let source = string(provenance, "source").ok_or_else(|| fail(message, "missing source"))?;
    if metadata.get("source").and_then(Value::as_str) != Some(source) {
        return Err(fail(message, "source attribution mismatch"));
    }
    let signature = object(provenance.get("signature"))
        .ok_or_else(|| fail(message, "missing signature state"))?;
    if signature.contains_key("scheme") {
        verify_signature(message, provenance, signature, source)
    } else {
        verify_attribution(message, signature, source)
    }
}

fn verify_attribution(
    message: &Message,
    signature: &Map<String, Value>,
    source: &str,
) -> Result<MessageAssurance, MessageVerificationError> {
    let status = string(signature, "status").ok_or_else(|| fail(message, "missing signature status"))?;
    let expected = match source {
        "discord" | "telegram" => "unsupported",
        "api" => "not_configured",
        "system" => "not_applicable",
        _ => return Err(fail(message, "unsigned native source")),
    };
    if status != expected {
        return Err(fail(message, "invalid signature status for source"));
    }
    Ok(MessageAssurance::Attributed {
        source: source.to_owned(),
        status: status.to_owned(),
    })
}

fn verify_signature(
    message: &Message,
    provenance: &Map<String, Value>,
    signature: &Map<String, Value>,
    source: &str,
) -> Result<MessageAssurance, MessageVerificationError> {
    if source != "ios" && source != "macos" {
        return Err(fail(message, "signed source is not native"));
    }
    if string(signature, "scheme") != Some(SIGNATURE_SCHEME) {
        return Err(fail(message, "unsupported signature scheme"));
    }
    let key_id = required(message, signature, "key_id")?;
    let public_key = decode_required(message, signature, "public_key")?;
    let claimed_key_id = decode(key_id).map_err(|detail| fail(message, detail))?;
    if claimed_key_id != digest(&SHA256, &public_key).as_ref() {
        return Err(fail(message, "signing key identifier mismatch"));
    }
    let compact = required(message, signature, "signed_message")?;
    let (header, payload, input, raw_signature) = parse_jws(message, compact)?;
    validate_header(message, &header, key_id)?;
    UnparsedPublicKey::new(&ECDSA_P256_SHA256_FIXED, public_key)
        .verify(input.as_bytes(), &raw_signature)
        .map_err(|_| fail(message, "invalid ES256 signature"))?;
    validate_payload(message, provenance, &payload)?;
    Ok(MessageAssurance::Verified {
        source: source.to_owned(),
        key_id: key_id.to_owned(),
    })
}

fn parse_jws(
    message: &Message,
    compact: &str,
) -> Result<(JwsHeader, SignedPayload, String, Vec<u8>), MessageVerificationError> {
    let segments: Vec<&str> = compact.split('.').collect();
    if segments.len() != 3 || segments.iter().any(|segment| segment.is_empty()) {
        return Err(fail(message, "malformed compact JWS"));
    }
    let header = decode_json(message, segments[0], "header")?;
    let payload = decode_json(message, segments[1], "payload")?;
    let signature = decode(segments[2]).map_err(|detail| fail(message, detail))?;
    if signature.len() != 64 {
        return Err(fail(message, "invalid ES256 signature length"));
    }
    Ok((header, payload, format!("{}.{}", segments[0], segments[1]), signature))
}

fn validate_header(
    message: &Message,
    header: &JwsHeader,
    key_id: &str,
) -> Result<(), MessageVerificationError> {
    if header.alg != "ES256" || header.typ != JWS_TYPE || header.kid != key_id {
        return Err(fail(message, "JWS header mismatch"));
    }
    Ok(())
}

fn validate_payload(
    message: &Message,
    provenance: &Map<String, Value>,
    payload: &SignedPayload,
) -> Result<(), MessageVerificationError> {
    if payload.version != 1 || payload.purpose != MESSAGE_PURPOSE || payload.action.kind != "reply" {
        return Err(fail(message, "signed payload contract mismatch"));
    }
    if payload.message_id.is_empty() || payload.issued_at <= 0 {
        return Err(fail(message, "invalid signed message identity"));
    }
    if message.body.as_deref() != Some(payload.body.as_str()) {
        return Err(fail(message, "signed body mismatch"));
    }
    if message.reply_to.as_deref() != Some(payload.action.message_id.as_str()) {
        return Err(fail(message, "signed reply target mismatch"));
    }
    let actor_id = object(provenance.get("actor")).and_then(|actor| string(actor, "id"));
    if actor_id != Some(payload.boss_id.as_str()) {
        return Err(fail(message, "signed boss identity mismatch"));
    }
    Ok(())
}

fn required<'a>(
    message: &Message,
    object: &'a Map<String, Value>,
    key: &str,
) -> Result<&'a str, MessageVerificationError> {
    string(object, key).ok_or_else(|| fail(message, &format!("missing {key}")))
}

fn decode_required(
    message: &Message,
    object: &Map<String, Value>,
    key: &str,
) -> Result<Vec<u8>, MessageVerificationError> {
    let encoded = required(message, object, key)?;
    decode(encoded).map_err(|detail| fail(message, detail))
}

fn decode_json<T: for<'de> Deserialize<'de>>(
    message: &Message,
    encoded: &str,
    part: &str,
) -> Result<T, MessageVerificationError> {
    let bytes = decode(encoded).map_err(|detail| fail(message, detail))?;
    serde_json::from_slice(&bytes).map_err(|_| fail(message, &format!("invalid JWS {part}")))
}

fn decode(value: &str) -> Result<Vec<u8>, &'static str> {
    if value.is_empty() || value.len() % 4 == 1 {
        return Err("invalid base64url encoding");
    }
    let mut output = Vec::with_capacity(value.len() * 3 / 4);
    let mut accumulator = 0_u32;
    let mut bits = 0_u8;
    for byte in value.bytes() {
        accumulator = (accumulator << 6) | u32::from(base64_value(byte)?);
        bits += 6;
        if bits >= 8 {
            bits -= 8;
            output.push((accumulator >> bits) as u8);
            accumulator &= (1_u32 << bits) - 1;
        }
    }
    if accumulator != 0 {
        return Err("non-canonical base64url encoding");
    }
    Ok(output)
}

fn base64_value(byte: u8) -> Result<u8, &'static str> {
    match byte {
        b'A'..=b'Z' => Ok(byte - b'A'),
        b'a'..=b'z' => Ok(byte - b'a' + 26),
        b'0'..=b'9' => Ok(byte - b'0' + 52),
        b'-' => Ok(62),
        b'_' => Ok(63),
        _ => Err("invalid base64url character"),
    }
}

fn object(value: Option<&Value>) -> Option<&Map<String, Value>> {
    value?.as_object()
}

fn string<'a>(object: &'a Map<String, Value>, key: &str) -> Option<&'a str> {
    object.get(key)?.as_str().filter(|value| !value.is_empty())
}

fn fail(message: &Message, detail: &str) -> MessageVerificationError {
    MessageVerificationError {
        message_id: message.id.clone(),
        detail: detail.to_owned(),
    }
}
